golem-cli/src/command/component.rs

```diff
@@ -50,6 +53,25 @@ pub enum ComponentSubCommand<ProjectRef: clap::Args, ComponentRef: clap::Args> {
         non_interactive: bool,
     },
 
+    /// Create a new component, taking all parameters from the app manifest
+    AddWithManifest {
```

Comment:
I'm not sure what your approach to backwards compat is, so I added this and update as new commands for now.

Reply:
How about existing `Add` command with optional `app`?
Same goes for `Update` .

cc @vigoo  What's your take?

Reply:
For now I would use the current commands with optional flags for either:
 - explicitly specifying the manifest files
 - or to auto detect the manifest files to be used
 
 Later we can make the second default "true".  This way most our docs should be up to date, and we can keep adding the new flags, also I would expect that eventually every command that works with components will have the "auto detect".

Reply:
The main problem I see here is that the non-manifest version takes a required positional  argument for the component wasm file:

```
❯ ./target/debug/golem-cli component add --help
Creates a new component with a given name by uploading the component WASM

Usage: golem-cli component add [OPTIONS] --component-name <COMPONENT_NAME> <component-file>

Arguments:
  <component-file>  The WASM file to be used as a Golem component

Options:
  -c, --component-name <COMPONENT_NAME>
          Name of the newly created component
  -v, --verbose...
          Increase logging verbosity
      --ephemeral
          Create an Ephemeral component. If not specified, the command creates a Durable component
  -q, --quiet...
          Decrease logging verbosity
      --durable
          Create a Durable component. This is the default
  -F, --format <FORMAT>
          
  -y, --non-interactive
          Do not ask for confirmation for performing an update in case the component already exists
  -h, --help
          Print help
```

I think it would be quite confusing for users if that argument would be required or unused depending on whether a flag for the manifest file is provided.

Additionally, wasm-rpc (and the new commands too) try to search for a manifest file automatically if none was provided through the cli and fail if none was found. IMO we definitely don't want to introduce that behaviour in the existing add command as it would break user's flows.

Could you please provide some pointers how you want this to work?

Reply:
I don't think that making a positional argument optional is an uncommon thing, especially if it's controlled by a new flag (or existence of new config files). Also, while we are aiming in general for backward compatibility, in the case of cli and tooling it is expected that we will deprecate / refactor things in the coming releases, given the new app manifest approach. So we should not increase too much the number of soon to be deprecated commands, I would rather keep adding flags to existing thing, and maybe later deprecate them together (or flags of them)

So I would make the wasm file optional depending the flags, given it is not breaking anything out there, and I think having so is quite common in tools (e.g. taking input from stdin vs file vs a config which points to a file).

(It also a good question that why we have this positional arg in the first place, when mostly everything else is by named flags, but this is already out :) )

Reply:
How should it behave in regards to automatic discovery of manifests?
In wasm-rpc it always does so, failing the command if it cannot find one.

Should we be inconsistent with wasm-rpc here and only support explicitly passed manifests?

Reply:
I would probably do this fallback logic:
- if the component-file is specified then use no manifests at all
- if the component-file is not specified, check for manifest flags, if none, do auto discovery
- if both component-file and manifest flags are defined then show an error that they are mutually exclusive

Reply:
I'm struggling to do this one with clap. I think the correct approach to do this is:

* Have an arg_group with two nested arg groups for explicit component file version and manifest based version. This one is required, the two nested ones are optional.
* In the nested ones, set requires to false but explicitly requires all fields in the arggroup, i.e.
```
#[derive(clap::Args, Debug, Clone)]
#[group(required = true, multiple = true)]
pub struct ExplicitComponentOrManifestArgGroup {
    #[command(flatten)]
    explicit_component: Option<ExplicitComponentArgGroup>,

    #[command(flatten)]
    manifest: Option<ManifestArgGroup>,
}

#[derive(clap::Args, Debug, Clone)]
#[group(required = false, multiple = true, requires_all = vec!["component_file", "component-type-flag"])]
pub struct ExplicitComponentArgGroup {
    /// The WASM file to be used as a Golem component
    #[arg(value_name = "component-file", value_hint = clap::ValueHint::FilePath)]
    component_file: PathBufOrStdin, // TODO: validate exists

    /// The component type. If none specified, the command creates a Durable component.
    #[command(flatten)]
    component_type: ComponentTypeArg,
}

#[derive(clap::Args, Debug, Clone)]
#[group(required = false, multiple = true, requires = "app")]
pub struct ManifestArgGroup {
    /// List of application manifests. Can be specified multiple times
    #[arg(long, short)]
    app: Vec<PathBuf>,
}
```
This way, if any of the args in one of the nested groups is present, all of the other ones are also requires. The outer arggroup ensures that no two conflicting arguments are present.


There are two problems
* This currently doesn't work with the positional arguments. Clap assumes there is an extra positional argument, but I'm not sure where it's coming from.
* The help messages for this are really bad. We need to have some custom docstring for sure that explains all these mutually exlusive options to users